### ISSUE TITLE
Improve keyboard navigation

### ISSUE THREAD

**Issue by tnajdek** [zotero/web-library#218]
We should improve keyboard navigation in items table and collections tree.

* https://www.w3.org/TR/wai-aria-practices/#grid
* https://www.w3.org/TR/wai-aria-practices/#TreeView

- [x] Header
- [x] Libraries/Collection Tree
- [x] Tag List
- [x] Tag Selector Toolbar
- [x] Items Toolbar
- [x] Items Table
- [x] Item Details Tabs
    - [x] Info
    - [x] Notes
    - [x] Tags
    - [x] Attachments
    - [ ] Related
    - [ ] Top-level attachment editor
    - [ ] Top-level notes editor

**Comment by flachware** [comment]
@dstillman @tnajdek I tried to wrap my head around keyboard navigation and focus handling. Currently we highlight selected items in the items table with a blue background, no matter where the focus is. E.g. you select an item with the mouse, use key up/down to select other items, seems logical. Then you edit a title in the info pane, the focus is somewhere in that pane now, you can’t use the keyboard navigation in the items table anymore, but you can’t see the difference. An even more clear example would be the notes list, I guess we want to be able to use keyboard navigation there too? In this case you select a note with the mouse, then you can use key up/down to select other notes – but the items table still has the blue background.

So, I really can’t think of another solution than the one of the current client (and most other apps): when the items table does not have the focus we gray out the selection. The drawback being that this makes the UI less colorful – which has been my motivation to keep it blue in the first place (based on the feedback I got on earlier drafts).

Are we fine with a grayed out active state when the items table does not have the focus? I’m checking back because this will also have some side effects, e.g. the item type icons have to work on a new background etc. Any stroke of genius to overcome this issue in another way?

In case we agree on the approach to use grayed out selections to indicate the absence of focus, it might be worth a thought to not show an outline around the items table when it has the tab focus, but auto-select the first item instead, like in the client.

**Comment by dstillman** [comment]
Yes, it should be gray when it doesn't have focus. Selection and focus aren't the same, and it's important to convey when they diverge.

While less important, arguably it should be gray even when the browser itself loses focus. That's how desktop apps works, and browsers do it for form controls for the same reason — it's helpful to know what on your screen is going to take keyboard input.

**Comment by flachware** [comment]
Ok. If I’m not mistaken the selection will be gray inherently when the browser loses the focus, as the items table will lose the focus too in this case (once we tie the blue color to the items table focus state).

How do you feel about auto-selecting the first item on tab focus instead of an outline around the items table?

**Comment by dstillman** [comment]
It should definitely select the first item (unless an item is already selected, of course).

**Comment by flachware** [comment]
@tnajdek I prepared the focus/active styles in the items table. We would need the following changes:

- Autoselect first item when the items table gets the focus (unless an item is already selected)
- Item type icons and tag color icons should only switch to the 'active' symbol (I’ll rename that to focus) when the table has the focus.
- I think we should remove the tabindices from each table header cell and set just one tabindex on the entire row and allow arrow navigation inside the row.

**Comment by tnajdek** [comment]
> Item type icons and tag color icons should only switch to the 'active' symbol (I’ll rename that to focus) when the table has the focus.

I've noticed you've commented out logic picking different item type icons when active. I've added the "is focused" condition but left the entire bit commented out, so you should be re-enable whenever icons are ready.

**Comment by flachware** [comment]
@tnajdek let’s rethink the whole tab cycle. According to the WAI-ARIA standards for treeview, toolbar, grid, and tabs I think the tab cycle should look like this:

- Collection tree
- Tag selector
- Toolbar
- Item table
- Panel

A variant would be to split up the panel step into a step for tabs and tab pane.

Does that make sense to you?

**Comment by dstillman** [comment]
Why would the toolbar go there? In the client the focus starts in the search bar (the rest of the toolbar isn't currently keyboard-accessible) and goes to collections → tag selector (individual tags inaccessible, because there's not really a good alternative) → tag selector filter → items list → item fields.

The main difference here is that the rest of the toolbar buttons is presumably accessible, but that's all the more reason to not put them between collections and items. Having a search field between collections and items actually sort of makes sense in terms of workflow, but it doesn't make sense to have to tab through the whole toolbar just to navigate items after you've changed collections.

**Comment by tnajdek** [comment]
I think toolbar would be one stop with arrow keys used for selecting individual tools (as per https://www.w3.org/TR/wai-aria-practices/#toolbar) so tabbing though the toolbar would be one step. It makes sense to have that step just before the items pane as @flachware suggested, and I don't think a single tab to skip through toolbar would add that much overhead. I can see only two alternatives to that, neither of which I like: having toolbar not accessible via keyboard at all or have it accessible out of visual order.

I think Tag selector would also be one stop with arrow keys for selecting individual tags (behaving like a grid).

**Comment by dstillman** [comment]
Oh, I forgot that the toolbar here doesn't extend above the collections pane as it does in the client. That order makes sense, then.

What about the search field? Seems like that's not positioned properly right now, but certainly you shouldn't have to Shift-Tab back through the top-right menu and account picture to filter the current view.

> I think toolbar would be one stop with arrow keys used for selecting individual tools

Ah, OK. I wasn't aware that that was an ARIA recommendation. That works for the web version. That's not really a native thing, though, so I don't think we'd want to do that for the toolbar in the client (though I think we could do it for the tag selector, which is a custom thing). At least on macOS, I think you need to use VoiceOver to access toolbar buttons, but I guess the idea is that you have equivalent menu options with keyboard shortcuts for all actions.

**Comment by flachware** [comment]
Tab cycle playground (feel free to edit):

- Navbar
- Collection tree
- Tag selector
- Toolbar
- Item table
- Panel

**Comment by flachware** [comment]
^ Each list item above would be a single tab stop. Navigation inside the components works with arrow keys.

Search is a good question. So first, the navbar would not consist of many tab stops but only have a single tab stop (https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-1/menubar-1.html). And then I think we are free to decide which element we 'focus' by default? Why not 'focus' on the search box when you tab to the navbar? And then it’s up to you to move to the left or to the right. This seems a bit similar to the items table, where we focus on the first item, but arrows let you move up to the table header.

**Comment by tnajdek** [comment]
I like this approach, 4 shift+tabs to get from items pane to the search box is not ideal but still reasonable and I can't think of an alternative that would be better than this.

I say we go ahead with this.

**Comment by flachware** [comment]
If I’m not mistaken you would need 5 shift+tabs in this case. The only reasonable (additional) alternative I can think of would be a shortcut, e.g. `/` like in other web apps (we could show all keyboard shortcuts when you hit `?`).

**Comment by tnajdek** [comment]
I've came up with a few more issues to consider while implementing keyboard navigation:

- How to enable navigation down to actions related to focused-element (e.g. "..." menu in collections tree or hover-buttons in attachments/notes)
- I've opted to use 2 stops for Tag Selector - one for tags list, one for the "toolbar", should we instead use one stop?
- Item details info tab contains list of fields, should this be navigable via up/down arrows for consistency with the rest of the UI or treated as a form (each field a separate tab stop)? Note: using arrow keys will interfere with select box which currently can be opened by pressing arrow down.

**Comment by flachware** [comment]
>If left/right arrows are used to navigate through header, this will interfere with search input where arrows are used to move cursor, how should we fix this?

The WAI-ARIA spec of the toolbar (which is similar enough I guess) component gives some clues that there is no perfect solution for this problem, we neither can’t avoid the input field nor does it help to put the input at the end of the menubar.

But having a look at the navigation menubar example (https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-1/menubar-1.html) it seems to me that left/right arrows are context-dependent here: when a top-level menu item or a sub item without submenu has the focus and you hit left/right, the focus moves in the menubar. However, when a sub item with a submenu has the focus, left/right keys will navigate inside the submenu – but one more keystroke in the same direction will move the focus again.

Could we do something similar, like, when there is no text inside the input left/right moves the focus and when there is text left/right moves the cursor inside the text until the beginning/end of the text and one more keystroke moves the focus?

In any case, when there is text in the search box and it has the focus, it should be possible to use left/right to move the cursor, right? So, let’s assume we already had this behavior, then left/right would just move the cursor inside the focused input and you would be trapped. A less-than-ideal approach would be to allow the user to move the focus by using up/down (as the suggested duplication of functionality in the toolbar spec), but it would be less intuitive than a context-sensitive solution.

**Comment by flachware** [comment]
>How to enable navigation down to actions related to focused-element (e.g. "..." menu in collections tree or hover-buttons in attachments/notes)?

I think right key makes most sense here. E.g. in collection tree, first right keystroke opens a collection (in case it has sub collections), second keystroke sets the focus on the '...' dropdown toggle. Does anything speak against that approach?

**Comment by flachware** [comment]
> I've opted to use 2 stops for Tag Selector - one for tags list, one for the "toolbar", should we instead use one stop?

I tend to say it should be one tab stop, but let’s keep it like that until we are working on the UI of the tag selector, agreed?

**Comment by flachware** [comment]
>Item details info tab contains list of fields, should this be navigable via up/down arrows for consistency with the rest of the UI or treated as a form (each field a separate tab stop)? Note: using arrow keys will interfere with select box which currently can be opened by pressing arrow down.

Could we solve this similar to Tropy? Like this (info pane): the entire panel has just one tab stop; when the panel has the focus, left and right keys move the focus in the tab list; Down arrow sets the focus on the item type selector; from then on, you are in 'form mode' and use tab to step through the editables – like it works now; but when you hit ESC and then tab you do not continue to step through the editables but the focus moves to the next component in the tab cycle (or the browser tabs in our case).

So basically we have to give the focus back to the panel when you hit ESC. Note that I’m not fully clear wether we have to set the focus on the panel or on the tab list inside.

**Comment by dstillman** [comment]
> So basically we have to give the focus back to the panel when you hit ESC or enter.

In the client, Enter/Esc in an item field refocuses the item in the middle pane, since that seems like the most likely next step if you're just going down your items list fixing metadata.

**Comment by flachware** [comment]
That makes sense to me, let’s stick to the client behavior.

**Comment by tnajdek** [comment]
Thanks for all the input, I'll start continue with the implementation based on your suggestions and come back if I have any further questions.

**Comment by flachware** [comment]
@tnajdek some navbar issues / enhancements:

- When you click into the search box and hit arrow left, the search options button has the focus, but you can’t see the focus because the global `.keyboard` class is not set. Could we set the keyboard class on arrow keys too generally?

- I think it would make sense to store the last focused element in the navbar like in the WAI-ARIA examples. E.g. when you leave the navbar and the search box had the focus and you come back the search box would be the first elements that gets the focus (instead of the Zotero logo).

- Does anything speak against 'looping' the focus in the navbar and toolbar?

- Another optimization would be to focus on the search box by default instead of the Zotero logo. This appears a bit non-standard to me, but it I think it would be in the best interest of the users, who will use the search feature more than anything in the navbar.

**Comment by flachware** [comment]
Now that we move the active class in the items list with arrow keys I think we should also move the `.selected` class in the desktop version of the collection-tree – and not just the `.focus` class. This would also be in line with the client.

**Comment by flachware** [comment]
@tnajdek keyboard class and looping focus look good to me!

Re nav-toggle and mobile-nav I think we could strictly stick to the menu button design pattern: https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-links.html. Except that we don’t want to show the menu on hover but on click.

**Comment by flachware** [comment]
@tnajdek would you mind if I close this issue in favour of individual keyboard navigation issues?

**Comment by tnajdek** [comment]
Sure, no problem
